alert-common/src/main/java/com/synopsys/integration/alert/common/workflow/processor/DigestMessageContentProcessor.java

```diff
@@ -51,41 +53,35 @@ public DigestMessageContentProcessor(final MessageContentCollapser messageConten
     }
 
     @Override
-    public List<AggregateMessageContent> process(final List<AggregateMessageContent> messages) {
+    public List<MessageContentGroup> process(final List<AggregateMessageContent> messages) {
         final List<AggregateMessageContent> collapsedMessages = messageContentCollapser.collapse(messages);
-        final Map<LinkableItem, List<AggregateMessageContent>> topicInfoToTopics = groupByTopic(collapsedMessages);
+        final Map<Pair<LinkableItem, LinkableItem>, List<AggregateMessageContent>> messagesGroupedByTopicAndSubTopic = groupByTopicAndSubTopicValue(collapsedMessages);
 
-        final List<AggregateMessageContent> digestMessages = new ArrayList<>();
-        for (final Map.Entry<LinkableItem, List<AggregateMessageContent>> topicInfoToTopicsEntry : topicInfoToTopics.entrySet()) {
-            final LinkableItem topicInfo = topicInfoToTopicsEntry.getKey();
-            final List<AggregateMessageContent> groupedMessages = topicInfoToTopicsEntry.getValue();
+        final Map<String, MessageContentGroup> messageGroups = new LinkedHashMap<>();
+        for (final Map.Entry<Pair<LinkableItem, LinkableItem>, List<AggregateMessageContent>> groupedMessageEntry : messagesGroupedByTopicAndSubTopic.entrySet()) {
+            final Pair<LinkableItem, LinkableItem> topicAndSubTopic = groupedMessageEntry.getKey();
+            final LinkableItem topic = topicAndSubTopic.getLeft();
+            final SortedSet<CategoryItem> combinedCategoryItems = gatherCategoryItems(groupedMessageEntry.getValue());
 
-            final Set<LinkableItem> newSubTopics = gatherSubTopics(groupedMessages);
-            final SortedSet<CategoryItem> newCategoryItems = gatherCategoryItems(groupedMessages);
-
-            final AggregateMessageContent newMessage = new AggregateMessageContent(topicInfo.getName(), topicInfo.getValue(), topicInfo.getUrl().orElse(null), newSubTopics, newCategoryItems);
-            digestMessages.add(newMessage);
+            final AggregateMessageContent newMessage = new AggregateMessageContent(topic.getName(), topic.getValue(), topic.getUrl().orElse(null), topicAndSubTopic.getRight(), combinedCategoryItems);
+            messageGroups.computeIfAbsent(topic.getValue(), ignored -> new MessageContentGroup()).add(newMessage);
```

Comment:
I mentioned it on the last PR and I think it's even more true with the most recent changes, We should create an object for the map being used here. Create a list of these objects and iterate through them. Add some methods to this object that does the above functionality. It will be easier to understand imo.